text/0018-open-context-provider-framework.md

```diff
@@ -0,0 +1,157 @@
+    1. Define the `identifier` field with appropriate value from `cxschema.ContextProvider` enum.
+    2. Add a call to `registerContextProvider()` to the bottom of each definition file.
+        * ex: In `context-provider/vpc.ts` we add the call 
+      `registerContextProvider(VpcNetworkContextProviderPlugin.identifier, VpcNetworkContextProviderPlugin);`
+3. Modify the definition of `availableContextProviders` in `context-providers/index.ts` to be initially empty.
+4. Modify the implementation of `registerContextProvider` in `context-providers/index.ts` to throw an error if trying to re-register a name.
+5. Use the existing `PluginHost` mechanism to import/register ; extend `PluginHost` with a `registerContextProviderSource()` method.
+6. In `registerContextProviderSource()` we:
+    1. Check that the given module implements `interface ContextProviderPlugin` as best we can, and throw an error if it does not.
+    2. Check whether the plugin has already been registered. If it has, then should we throw an error here, or a warning, or... ?
+    3. Invoke `registerContextProvider()` to register the plugin
+
+### In CDK modules
+
+A saving grace here for the Construct API is that even though `enum ContextProvider` in `@aws-cdk/cloud-assembly-schema/lib/context-queries.ts`
+is a static enum that defines the names of each of the built-in `ContextProvider`s, the `interface GetContextKeyOptions` defines the `provider`
+field as being a string. So, we shouldn't require any API changes here.
+
+A challenge that I can see is that the call to `stack.reportMissingContext()` in `ContextProvider.getValue()` casts the `provider` value to a
+`cxschema.ContextProvider` (i.e. to the hard-coded enum), and then the call-chain under that requires that the `provider` be of the
+`cxschema.ContextProvider` type. For a plugable `ContextProvider` to work, all of the `providers` in this call-chain would have to have their
+type changed to `string`. I don't *think* that this will negatively impact the `CloudAssembly` -- the value of the `provider` is a string in the
+`cdk.out` json file, so changing it to a string internally should not impact that.
+
+# Drawbacks
+
+This change would alter the internal API in that some fields that are currently defined as `cxschema.ContextProvider` would need to change to
+`string` type. In practice, this means that the `ContextProvider` plugins would have a hard requirement of requiring at minimum a specific
+version of the `@aws-cdk/cloud-assembly-schema` and `@aws-cdk/core` modules.
+
+# Rationale and Alternatives
+
+Other options may be possible. This is one proposal by someone not intimately familiar with the internals of the CDK toolchain.
+
+# Adoption Strategy
+
+This is a breaking change. Adoption would require updating both the CDK CLI and
+CDK construct library that are in use.
+
+# Unresolved questions
+
+* Is there a precidence for coupling a minumum version of the CDK CLI with the a minimum version of CDK construct library?
```

Comment:
Oh yes, we do this all the time. Our stated compatibility rules at the moment are CLI >= framework.